drivers/tty/serdev/pibridge.c

```diff
@@ -116,15 +116,15 @@ static void pibridge_remove(struct serdev_device *serdev)
 
 #ifdef CONFIG_OF
 static const struct of_device_id pibridge_of_match[] = {
-	{ .compatible = "kunbus,pibridge" },
+	{ .compatible = "kunbus,pibridge-comm" },
```

Comment:
I think this is wrong. Handling of the sniff pins will have to be moved to this driver so that it can perform the bus scan and instantiate devices on the PiBridge bus. This is really "the" PiBridge driver. If piControl toggles the sniff pins for now, rename the node to which the piControl platform device driver binds.

The only other thing that's done on the PiBridge is the Gateway communication but that's completely separate.

Reply:
Huh? IIRC our plan was to implement a pibridge bus driver. Device detection is definitely a job for a bus driver. The bus driver then uses the serdev for communication but nothing else...

Reply:
There seems to be a misunderstanding. This *is* the pibridge bus driver, or the first portion of it.

For now, this driver only provides a set of functions to perform RS-485 communication. There's still some work to do after this pull request is merged to convert all functions in piControl over to the new RS-485 communication primitives.

The next step will be to add bus scan functionality to the driver (using the sniff pins). We'll then define a `struct pibridge_dev` (which embeds a `struct device`) and instantiate one for each device found. That will allow us to change the communication primitives such that they're no longer invoked with a bus address parameter (an integer), but with a `struct pibridge_dev` parameter. The struct contains the bus address among other things.

As a third step, we can define individual device types for DIO, AIO, MIO and assign that type to the `struct device` which is instantiated for each found PiBridge device. That will allow us to write drivers for each module which bind to the device of same type.

To summarize, what this driver will be doing is:
* scan the bus using the sniff pins, instantiate devices
* provide RS-485 communication primitives which are used by the devices' drivers

Reply:
AFAIU for a bus driver we have to define a `bus_type` and then register it with the `bus_register()`. Will this be done in this file (pibridge.c) then?

Reply:
I think my understanding was: We have a kernel module which defines a `bus_type` "pibridge" (or something like that), which we register with `bus_register()`. Then in the `probe()` function this module discovers all physically attached devices by using the sniff pins. For each found device as struct `device` is instantiated and registered with `device_register()` (this fills the `klist_devices` member of the struct `subsys_private` member of `bus_type`). 
As soon as a driver for one (or more if of the same type) of the devices is registered `bus_type->match()` is called and checks the registered driver against the list of found devices (a match fills the list of `klist_drivers` in the `subsys_private` struct). If a match is found the `probe()` function of the device driver for the specific device is called (done by the device core). In that function the device driver sends its configuration data via RS485. For this it uses the library functions provided in pibridge.c (which for that reason better should be called pibridge-comm.c and which is the reason why I chose "kunbus,pibridge-comm" as the compatible name). So in my understanding the pibride communiction module (aka pibridge.c) and the bus driver module are different modules...

Reply:
<blockquote>
AFAIU for a bus driver we have to define a bus_type and then register it with bus_register(). Will this be done in this file (pibridge.c) then?
</blockquote>

Yes. It could be in a separate file if you think that it becomes too large but that other file and `pibridge.c` would be linked together to form a driver which performs the RS-485 communication and provides a bus on top of it. Including device enumeration on that bus using the sniff pins. It's all part of that single driver. So the sniff pins handling would also be moved here eventually.

In the device hierarchy, there's the UART device, plus the pibridge platform device below the UART, plus the bus below the pibridge platform device, plus all the module devices on that bus.

Reply:
You can call the *file* `pibridge-comm.c`, but  not the compatible string. There are not two separate platform devices in the device hierarchy (one for the RS-485 communication and one for the bus on top of it). At least I couldn't see a benefit of such a separation.

Reply:
What about the approach I sketched above (with a pibridge-comm library module and a bus driver module)?

Reply:
Only makes sense if you intend to let the bus driver modul run on top of some other communication protocol. (PiBridge 2.0?) Otherwise just adds overhead and confusion IMO.